File: coding-challenges/sources/leetcode/811. Subdomain Visit Count/811-SubdomainVisitCount.swift


fileprivate func subdomainVisitCount(_ cpdomains: [String]) -> [String] {
    // Create an empty dictionary to store the subdomain counts
    var subdomainCounts = [String: Int]()

    // Iterate over each input domain
    for cpdomain in cpdomains {
        // Split the input domain into its count and domain components
        let components = cpdomain.split(separator: " ")
        let count = Int(components[0])!
        var subdomain = String(components[1])

        // Iterate over each subdomain in the domain
        while !subdomain.isEmpty {
            // Add the count to the subdomain count in the dictionary
            subdomainCounts[subdomain, default: 0] += count

            // Remove the current subdomain from the domain and continue with the next subdomain
            if let range = subdomain.range(of: ".") {
                subdomain = String(subdomain[range.upperBound...])
            } else {
                subdomain = ""
            }
        }
    }

    // Convert the dictionary to an array of strings in the expected format
    return subdomainCounts.map { "\($0.value) \($0.key)" }
}

import XCTest

final class SubdomainVisitCountTests: XCTestCase {
    func test() {
        var cpdomains = ["9001 discuss.leetcode.com"]
        var expected = Set(["9001 discuss.leetcode.com", "9001 leetcode.com", "9001 com"])
        XCTAssertEqual(Set(subdomainVisitCount(cpdomains)), expected)

        cpdomains = ["900 google.mail.com", "50 yahoo.com", "1 intel.mail.com", "5 wiki.org"]
        expected = Set(["901 mail.com","50 yahoo.com","900 google.mail.com","5 wiki.org","5 org","1 intel.mail.com","951 com"])
        XCTAssertEqual(Set(subdomainVisitCount(cpdomains)), expected)
    }
}

File: coding-challenges/sources/leetcode/811. Subdomain Visit Count/811-SubdomainVisitCount.py
def subdomainVisitCount(cpdomains):
    subdomain_counts = {}
    for cpdomain in cpdomains:
        count, domain = cpdomain.split()
        count = int(count)
        subdomains = domain.split('.')
        subdomain = ""
        for i in range(len(subdomains)-1, -1, -1):
            subdomain = subdomains[i] + ('.' if subdomain else '') + subdomain
            subdomain_counts[subdomain] = subdomain_counts.get(subdomain, 0) + count
    return [f"{count} {subdomain}" for subdomain, count in subdomain_counts.items()]

import unittest

class TestSubdomainVisitCount(unittest.TestCase):
    def test(self):
        self.assertSetEqual(set(subdomainVisitCount(["9001 discuss.leetcode.com"])), set(["9001 discuss.leetcode.com", "9001 leetcode.com", "9001 com"]))
        self.assertSetEqual(set(subdomainVisitCount(["900 google.mail.com", "50 yahoo.com", "1 intel.mail.com", "5 wiki.org"])), set(["901 mail.com","50 yahoo.com","900 google.mail.com","5 wiki.org","5 org","1 intel.mail.com","951 com"]))

if __name__ == '__main__':
    unittest.main()